lib/vtls/openssl.c

```diff
@@ -2672,6 +2691,11 @@ static CURLcode ossl_connect_step1(struct Curl_easy *data,
     ctx_options &= ~SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS;
 #endif
 
+  /* As OpenSSL may disable protocol versions by default, clear these
+     options first.
```

Comment:
Are you sure about this? I don't see it documented that way, I checked https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_clear_options.html

Reply:
Also, it seems like an unrelated change. It is not here to "suppress deprecation warnings" surely?

What's the reason for adding this function call? Everything that changes the setup of OpenSSL creates a risk and we need to be careful.

Reply:
> Are you sure about this?

@jay : On the page you linked:

> SSL_CTX_set_options() adds the options set via bitmask in options to ctx. **Options already set before are not cleared!**

and

> SSL_OP_NO_SSLv2
>     Do not use the SSLv2 protocol. **As of OpenSSL 1.0.2g the SSL_OP_NO_SSLv2 option is set by default.**

@bagder :
> Also, it seems like an unrelated change. It is not here to "suppress deprecation warnings" surely?

It is related because when conditionals enable `TLS_clent_method()`, there's a risk that the target protocol version is disabled by default. The documentation does not specify if setting the min/max version in the protocol overrides the options or not.

OpenSSL changed the handling of protocol selection many times in the several last versions: they deprecated `*_client_method()`, set `SSL_OP_NO_SSLv2` by default, then defined the later as `0`. I think clearing the protocol options that are under our control is the simplest way of dealing with the problem for all versions up to now.

Reply:
BTW: there's nothing in our test environment that allows to verify the proper TLS protocol and cipher selection. I checked the code to see if this can be implemented easily: this answer is no :-( Server side uses stunnel and no backend-independent API allows it on the client side.